### ISSUE TITLE
Networking issue with Swarm Mode on Windows 10 15063

### ISSUE THREAD

**Issue by jcorioland** [docker/docker#32326]
Hi,

On Windows 10 15063, I can not contact the container on the NAT ip from the inside or the outside.

Steps-to-reproduce:

- Enable Containers and Hyper-V features
- Reboot
- Install Docker Engine and Docker Service manually following https://docs.microsoft.com/en-us/virtualization/windowscontainers/manage-docker/configure-docker-daemon or using Docker for Windows (same behaviors)
- Reboot
- Pull microsoft/nanoserver:latest
- Create an overlay network: `docker network create -d overlay --opt com.docker.network.bridge.enable_icc=true simplechat`
- Create a service : `docker service create --name=testnano --endpoint-mode dnsrr --network testoverlay microsoft/nanoserver ping -t localhost`
- Scale the service to two containers: `docker service scale testnano=2`
- Enter powershell session in both containers using : `docker exec -it ID_CONTAINER powershell`
- Get the IP address on the nat network  in both container
- Container A cannot ping Container B
- Container B cannot ping Container A
- Both containers A & B can ping the gateway

The box has been migrated from RS1 to RS2 (insider slow).

Julien

**Comment by dineshgovindasamy** [comment]
Me @msabansal  are looking into this issue

**Comment by sixeyed** [comment]
Confirmed this is also a problem with Windows Server 2016 and the latest KB4015217 overlay patch. 

Using [this script](https://gist.github.com/sixeyed/3738d6f609db295f7a3866234669d031) you can initialize a single-node swarm, create an overlay network, and deploy an IIS service publishing port 80 to the host.

When the service is running you cannot access it on the host. Externally the IIS site is available, but on the server you can't access it on `localhost` (expected) or the container's internal IP - e.g. `10.0.0.2`.

**Comment by msabansal** [comment]
@sixeyed  This is expected. You can do container<->container traffic but not host<->container. You can do it through routing through the gateway only when the container is not connected to an overlay network. When the container is connected to overlay network we disable this. What @jcorioland  was facing was another issue which we resolved.

**Comment by sixeyed** [comment]
@msabansal thanks for explaining - will that be a permanent restriction? As a dev I can switch to swarm mode on my laptop and run my app using the exact same platform as test and prod. But if I can't access the containers, I can't verify the app. In Linux you can use loopback to access services on `localhost`, but that's not an option in Windows.

**Comment by rehmana1** [comment]
@msabansal So, if i want to access my swarm service (I have about 20 containers running in that service and all on a single Manager node) from my host machine (Which is Windows 10) how is it achievable and how will the traffic be redirected in between  the containers?

Current network is as follows:
Network - Ingress :: Scope - Swarm :: Driver - Overlay

**Comment by msabansal** [comment]
Container to container communication will work as expected. Every overlay deployment creates a gateway interface on the host that can be viewed when you execute "ipconfig /showallcompartments". You should be able to communicate using that compartment using ctstraffic: http://ctstraffic.codeplex.com/

Example: 
 C:\tools\ctstraffic -target:10.255.0.5 -port:80 -protocol:tcp  -compartment:4

4 is the compartment of the networks' gateway that was created on the host

**Comment by friism** [comment]
@rehmana1 To add to Sandeep's response: If you need to access a container running a service from the same host, you should use the container IP. For example:

```
PS C:\code\azure-template> docker service ls
ID                  NAME                MODE                REPLICAS            IMAGE                  PORTS
k3ddjb9jkrny        laughing_pare       replicated          1/1                 microsoft/iis:latest
PS C:\code\azure-template> docker ps
CONTAINER ID        IMAGE                  COMMAND                   CREATED             STATUS              PORTS               NAMES
2d37f228cb66        microsoft/iis:latest   "C:\\ServiceMonitor..."   2 minutes ago       Up About a minute   80/tcp              laughing_pare.1.nxl3d8ss8lp97p9etw5dxfoyq
PS C:\code\azure-template> docker inspect --format "{{.NetworkSettings.Networks.nat.IPAddress}}" 2d
172.26.161.69
PS C:\code\azure-template>
PS C:\code\azure-template>
PS C:\code\azure-template> docker service ls
ID                  NAME                MODE                REPLICAS            IMAGE                  PORTS
k3ddjb9jkrny        laughing_pare       replicated          1/1                 microsoft/iis:latest
PS C:\code\azure-template> docker ps
CONTAINER ID        IMAGE                  COMMAND                   CREATED             STATUS              PORTS               NAMES
2d37f228cb66        microsoft/iis:latest   "C:\\ServiceMonitor..."   4 minutes ago       Up 4 minutes        80/tcp              laughing_pare.1.nxl3d8ss8lp97p9etw5dxfoyq
PS C:\code\azure-template> docker inspect --format "{{.NetworkSettings.Networks.nat.IPAddress}}" 2d
172.26.161.69
PS C:\code\azure-template> iwr -UseBasicParsing 172.26.161.69


StatusCode        : 200
StatusDescription : OK
Content           : <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
                    <html xmlns="http://www.w3.org/1999/xhtml">
                    <head>
                    <meta http-equiv="Content-Type" cont...
RawContent        : HTTP/1.1 200 OK
                    Accept-Ranges: bytes
                    Content-Length: 703
                    Content-Type: text/html
                    Date: Wed, 05 Jul 2017 19:48:01 GMT
                    ETag: "1fc7c93fb6e4d21:0"
                    Last-Modified: Wed, 14 Jun 2017 02:30:55 GMT
                    Serve...
Forms             :
Headers           : {[Accept-Ranges, bytes], [Content-Length, 703], [Content-Type, text/html], [Date, Wed, 05 Jul 2017 19:48:01 GMT]...}
Images            : {@{outerHTML=<img src="iisstart.png" alt="IIS" width="960" height="600" />; tagName=IMG; src=iisstart.png; alt=IIS; width=960; height=600}}
InputFields       : {}
Links             : {@{outerHTML=<a href="http://go.microsoft.com/fwlink/?linkid=66138&amp;clcid=0x409"><img src="iisstart.png" alt="IIS" width="960" height="600" /></a>; tagName=A; href=http://go.microsoft.com/fwlink/?linkid=66138&amp;clcid=0x409}}
ParsedHtml        :
RawContentLength  : 703

```

**Comment by rehmana1** [comment]
@friism & @msabansal  .. Thanks for the information provided.
Sorry, i was not clear in explaining my requirement. What i need is to access the containers from any machine which is in the same network as my host machine.

I tried exposing the host port during service creation:
 docker service create --publish 5353:80 --replicas 5 --name democtr --update-delay 30s --update-parallelism 1 demoimage1 ping docker.com

but in this case the service is created with zero containers. Is there something missing in my command?
Also if I create the service without the --publish <PORT> option its creating the service and the containers.

**Comment by friism** [comment]
@rehmana1 you can use host-mode publish. That way, ports are published directly on hosts running tasks for the service:

```
docker service create --publish mode=host,target=80,published=80 microsoft/iis
```

**Comment by msabansal** [comment]
@friism Adding target to publishing mode will only allow 1 service instance per host in the cluster though.

**Comment by friism** [comment]
@msabansal yes, but I don't really see why you'd want to run many instances of the same service on a single host anyway - it's more efficient to just run one instance per host with a bigger resource allocation.
